torch/_dynamo/variables/misc.py

```diff
@@ -234,11 +234,44 @@ def visit(node):
 
         VariableTracker.apply(visit, (args, kwargs))
 
+        ctx = AutogradFunctionContextVariable.create(tx)
+        args = [ctx, *args]
+
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            from .torch import is_fn_safe_to_run, TorchHigherOrderOperator
+
+            def trampoline_autograd_apply(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            def trampoline_autograd_fwd(*args, **kwargs):
+                return self.fn_cls.forward(*args, **kwargs)
+
+            def trampoline_autograd_bwd(*args, **kwargs):
+                return self.fn_cls.backward(*args, **kwargs)
+
+            # Speculate fwd, will raise unimplemented and bubble up if not sound, or will accumulate guards
+            # onto tx if sound.
+            # TODO(voz): NOTE: This function identity is unguarded, but the odds of someone swapping self.fn_cls from
+            # autograd fn to something else is very low.
+            higher_order_autograd_fn = TorchHigherOrderOperator(trampoline_autograd_fwd)
+            speculated_fwd_result = higher_order_autograd_fn.call_function(
+                tx, args, kwargs
+            )
+            bwd_args = [ctx, speculated_fwd_result]
```

Comment:
To paraphrase @ezyang - "What about varying strides of grads?". This is an edge case we don't handle here today, but seems moderately safe to ignore for now? Esp since we don't use the result of backward for anything?

Reply:
It does matter if the autograd.Function branches on the strides of the Tensor and is actually unsafe in one of the branches (maybe there is a print statement or something because the user is debugging their code). The actual case where this matters is if a Tensor is a Tensor subclass -- it would be more likely that in the Tensor subclass case there will be unsafe behavior.

@Chillee's opinion is that this is unlikely to happen.

If we want to be conservative and make this safer, then... we could rewrite the backward function to coerce the grads to be the correct strides (or just speculate contiguous grads and then force the grads to be contiguous), and also have it assert that all the grads are plain old Tensors. We can do this by setting up a new autograd.Function that subclasses from the user's.

Happy to defer to the crowd on how much safety we need here.

Reply:
I am going to leave a comment and then punt on this.